### ISSUE TITLE
rlhf，grpo，qwen30B， Stuck in the initial stage of training

### ISSUE THREAD

**Issue by helin-wang-zte** [modelscope/ms-swift#5949]
**When training qwen3-14B and -32B with the same script, it can be completed smoothly. But switching to 30B won't work and won't generate any errors. Is it a problem with the model type 'qwen3_moe'?**
The script is as follows:
```
#rollout:
CUDA_VISIBLE_DEVICES=0,1,2,3 \
swift rollout \
    --model "$MODEL_PATH" \
    --model_type qwen3_moe \
    --vllm_tensor_parallel_size 2 \
    --vllm_data_parallel_size 2 \
    --port 8099
```

```
#traning:
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
NPROC_PER_NODE=8 \
swift rlhf \
    --rlhf_type grpo \
    --model "$MODEL_PATH" \
    --model_type qwen3_moe \
    --dataset /mnt/tenant-home_speed/whl/swift_learn/dataset/numina_math_64#32 \
    --output_dir /mnt/tenant-home_speed/whl/swift_learn/output/grpo-training/qwen30B \
    --importance_sampling_level sequence \
    --reward_funcs accuracy \
    --torch_dtype bfloat16 \
    --beta 0.0 \
    --epsilon 3e-4 \
    --epsilon_high 4e-4 \
    --steps_per_generation 4 \
    --num_train_epochs 1 \
    --per_device_train_batch_size 2 \
    --gradient_accumulation_steps 8 \
    --num_generations 16 \
    --train_type full \
    --use_vllm true \
    --vllm_mode server \
    --vllm_server_host xxxx \
    --vllm_server_port 8099 \
    --vllm_gpu_memory_utilization 0.6 \
    --vllm_max_model_len 4096 \
    --max_completion_length 2048 \
    --learning_rate 1e-6 \
    --save_total_limit 2 \
    --logging_steps 1 \
    --warmup_ratio 0.05 \
    --dataloader_num_workers 4 \
    --log_completions true \
    --deepspeed zero3 \
    --offload_optimizer true \
    --offload_model true \
    --sleep_level 1
```

<img width="1231" height="626" alt="Image" src="https://github.com/user-attachments/assets/21347236-fa47-4131-8a37-6e7f18428464" />

**Comment by genagent-oss** [comment]
same problem, have you fix this?

**Comment by hjh0119** [comment]
Training optimizations for MoE models will be completed soon

**Comment by helin-wang-zte** [comment]
> Training optimizations for MoE models will be completed soon

May I ask where I can stay informed in a timely manner after the optimization is completed?

**Comment by hjh0119** [comment]
We’ll update this issue once it’s done, stay tuned

**Comment by genagent-oss** [comment]
> We’ll update this issue once it’s done, stay tuned

 is the optimization done？

**Comment by genagent-oss** [comment]
> We’ll update this issue once it’s done, stay tuned

is this doen？

**Comment by hjh0119** [comment]
work in progress, will be completed within a few days.

https://github.com/modelscope/ms-swift/pull/5773
https://github.com/modelscope/ms-swift/pull/6025

**Comment by hjh0119** [comment]
weight synchronization is optimized in https://github.com/modelscope/ms-swift/pull/5773

please retry with main branch

some moe grpo script can be found in here https://github.com/modelscope/ms-swift/tree/main/examples/train/grpo

**Comment by helin-wang-zte** [comment]
> weight synchronization is optimized in [#5773](https://github.com/modelscope/ms-swift/pull/5773)
> 
> please retry with main branch
> 
> some moe grpo script can be found in here https://github.com/modelscope/ms-swift/tree/main/examples/train/grpo

```
MODEL_PATH="/mnt/tenant-home_speed/shared/models/huggingface/Qwen3-30B-A3B-Instruct-2507--e4dec04-C15"
MODEL_TYPE="qwen3_moe"

export WANDB_MODE=disabled
CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
NPROC_PER_NODE=8 \
swift rlhf \
    --rlhf_type grpo \
    --model "$MODEL_PATH" \
    --model_type "$MODEL_TYPE" \
    --dataset /mnt/tenant-home_speed/whl/swift_learn/dataset/code_gspo_1031/code_14057_gspo-3559_swift.json#64 \
    --output_dir /mnt/tenant-home_speed/whl/swift_learn/output/grpo-training/qwen30B \
    --external_plugins /mnt/tenant-home_speed/whl/swift_learn/grpo_plugin/plugin.py \
    --reward_funcs external_code1031_reward \
    --importance_sampling_level sequence \
    --use_vllm true \
    --vllm_mode colocate \
    --vllm_gpu_memory_utilization 0.4 \
    --vllm_tensor_parallel_size 2 \
    --vllm_max_model_len 25600 \
    --train_type full \
    --torch_dtype bfloat16 \
    --max_length 12000 \
    --max_completion_length 24576 \
    --overlong_filter true \
    --num_train_epochs 1 \
    --per_device_train_batch_size 1 \
    --learning_rate 1e-6 \
    --gradient_accumulation_steps 4 \
    --save_strategy 'steps' \
    --eval_strategy 'steps' \
    --eval_steps 1000 \
    --save_steps 1000 \
    --save_total_limit 10 \
    --logging_steps 1 \
    --warmup_ratio 0.01 \
    --dataloader_num_workers 4 \
    --num_generations 16 \
    --temperature 1.0 \
    --deepspeed zero3_offload \
    --log_completions true \
    --sleep_level 1 \
    --report_to tensorboard \
    --num_iterations 1 \
    --beta 0.001 \
    --move_model_batches 10
```

When training the 30b-moe-thinking model, an external reward function was used. Stuck after calling '__call__'. The utilization rate of 8 GPUs has always been 100%. The log, as shown in the figure, did indeed execute until the end of __call__().
在训练 30b-moe-thinking 模型时，使用了外部奖励函数。在调用“__call__”后卡死。8个GPU利用率一直是100%。日志如图所示，确实执行到了  __call__() 的最后。

<img width="746" height="437" alt="Image" src="https://github.com/user-attachments/assets/1d671c11-4c39-4731-83ef-69a8e1b20d45" />

<img width="945" height="433" alt="Image" src="https://github.com/user-attachments/assets/c9677f92-b8cc-43b9-9a7a-41155e3de233" />

**Comment by hjh0119** [comment]
Megatron GRPO is supported
